File: src/main/java/com/zhongan/devpilot/completions/inline/InlineCompletionCache.java
package com.zhongan.devpilot.completions.inline;

import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.util.Key;
import com.zhongan.devpilot.completions.prediction.DevPilotCompletion;

import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

public class InlineCompletionCache {
    public static final InlineCompletionCache INSTANCE = new InlineCompletionCache();

    private static final Key<List<DevPilotCompletion>> INLINE_COMPLETIONS_LAST_RESULT = Key.create("INLINE_COMPLETIONS_LAST_RESULT");

    private InlineCompletionCache() {
    }

    public static void store(Editor editor, List<DevPilotCompletion> completions) {
        editor.putUserData(INLINE_COMPLETIONS_LAST_RESULT, completions);
    }

    public static void clear(Editor editor) {
        editor.putUserData(INLINE_COMPLETIONS_LAST_RESULT, null);
    }

    public List<DevPilotCompletion> retrieveAdjustedCompletions(Editor editor, String userInput) {
        List<DevPilotCompletion> completions = editor.getUserData(INLINE_COMPLETIONS_LAST_RESULT);
        if (completions == null) {
            return Collections.emptyList();
        }
        return completions.stream()
            .filter(completion -> completion.getSuffix().startsWith(userInput))
            .map(completion -> completion.createAdjustedCompletion(
                completion.getOldPrefix() + userInput,
                completion.getCursorPrefix() + userInput))
            .collect(Collectors.toList());
    }
}

File: src/main/java/com/zhongan/devpilot/completions/common/inline/InlineCompletionCache.kt
package com.zhongan.devpilot.completions.common.inline

import com.intellij.openapi.editor.Editor
import com.intellij.openapi.util.Key
import com.zhongan.devpilot.completions.common.prediction.DevPilotCompletion
import java.util.stream.Collectors

class InlineCompletionCache {

    companion object {
        @JvmStatic
        val instance = InlineCompletionCache()

        private val INLINE_COMPLETIONS_LAST_RESULT = Key.create<List<DevPilotCompletion>>("INLINE_COMPLETIONS_LAST_RESULT")
    }

    fun store(editor: Editor, completions: List<DevPilotCompletion>) {
        editor.putUserData(INLINE_COMPLETIONS_LAST_RESULT, completions)
    }

    fun retrieveAdjustedCompletions(editor: Editor, userInput: String): List<DevPilotCompletion> {
        val completions = editor.getUserData(INLINE_COMPLETIONS_LAST_RESULT)
            ?: return emptyList()
        return completions.stream()
            .filter { completion: DevPilotCompletion -> completion.suffix.startsWith(userInput) }
            .map { completion: DevPilotCompletion ->
                completion.createAdjustedCompletion(
                    completion.oldPrefix + userInput,
                    completion.cursorPrefix + userInput
                )
            }
            .collect(Collectors.toList())
    }

    fun clear(editor: Editor) {
        editor.putUserData(INLINE_COMPLETIONS_LAST_RESULT, null)
    }
}
